dart/common/detail/AddonManager.h

```diff
@@ -0,0 +1,240 @@
+{
+  return (get<T>() != nullptr);
+}
+
+//==============================================================================
+template <class T>
+T* AddonManager::get()
+{
+  AddonMap::iterator it = mAddonMap.find( typeid(T) );
+  if(mAddonMap.end() == it)
+    return nullptr;
+
+  return static_cast<T*>(it->second.get());
+}
+
+//==============================================================================
+template <class T>
+const T* AddonManager::get() const
+{
+  return const_cast<AddonManager*>(this)->get<T>();
+}
+
+//==============================================================================
+template <class T>
+void AddonManager::set(const T* addon)
+{
+  mAddonMap[typeid(T)] = addon->cloneAddon(this);
+}
+
+//==============================================================================
+template <class T>
+void AddonManager::set(std::unique_ptr<T>&& addon)
+{
+  becomeManager(addon.get());
+  mAddonMap[typeid(T)] = std::move(addon);
+}
+
+//==============================================================================
+template <class T, typename ...Args>
+T* AddonManager::create(Args&&... args)
+{
+  T* addon = new T(this, std::forward<Args>(args)...);
+  mAddonMap[typeid(T)] = std::unique_ptr<T>(addon);
+
+  return addon;
+}
+
+//==============================================================================
+template <class T>
+void AddonManager::erase()
+{
+  AddonMap::iterator it = mAddonMap.find( typeid(T) );
+  DART_COMMON_CHECK_ILLEGAL_ADDON_ERASE(erase, it, DART_BLANK)
+  if(mAddonMap.end() != it)
+    it->second = nullptr;
+}
+
+//==============================================================================
+template <class T>
+std::unique_ptr<T> AddonManager::release()
+{
+  std::unique_ptr<T> extraction = nullptr;
+  AddonMap::iterator it = mAddonMap.find( typeid(T) );
+  DART_COMMON_CHECK_ILLEGAL_ADDON_ERASE(release, it, nullptr)
+  if(mAddonMap.end() != it)
+    extraction = std::unique_ptr<T>(static_cast<T*>(it->second.release()));
+
+  return extraction;
+}
+
+} // namespace common
+} // namespace dart
+
+//==============================================================================
+#define DART_ENABLE_ADDON_SPECIALIZATION()\
+  public:\
+  template <class T> bool has() const { return AddonManager::has<T>(); }\
+  template <class T> T* get() { return AddonManager::get<T>(); }\
+  template <class T> const T* get() const { return AddonManager::get<T>(); }\
+  template <class T> void set(const T* addon) { AddonManager::set<T>(addon); }\
+  template <class T> void set(std::unique_ptr<T>&& addon) { AddonManager::set<T>(std::move(addon)); }\
+  template <class T> void erase() { AddonManager::erase<T>(); }\
+  template <class T> std::unique_ptr<T> release() { return AddonManager::release<T>(); }\
+  template <class T, typename ...Args> T* create(Args&&... args)\
+  {\
+    T* addon = new T(this, std::forward<Args>(args)...);\
+    mAddonMap[typeid(T)] = std::unique_ptr<T>(addon); return addon;\
+  }
```

Comment:
This is actually a matter of C++ syntax. In order for a class that inherits ``AddonManager`` to specialize Addons for itself, it needs to **shadow** these templated functions before it can specialize these templated functions. Therefore, these templated functions need to be declared and defined in any derived class that wants specialized Addons.

One concern that just occurred to me is whether a derived class can benefit from the Addon specializations of a class that it inherits. This shadowing might prevent that from happening, I'm not sure. This is something I should test.